File: models/journal.go
package models

import (
	"encoding/json"
	"time"

	"github.com/gobuffalo/pop/v6"
	"github.com/gobuffalo/validate/v3"
	"github.com/gobuffalo/validate/v3/validators"
	"github.com/gofrs/uuid"
)

type Category string

const (
	Pests       Category = "Pests"
	Planting    Category = "Planting"
	Watering    Category = "Watering"
	Pruning     Category = "Pruning"
	Harvesting  Category = "Harvesting"
	Weather     Category = "Weather"
	Germination Category = "Germination"
)

// Journal is used by pop to map your journals database table to your go code.
type Journal struct {
	ID              uuid.UUID `json:"id" db:"id"`
	Title           string    `json:"title" db:"title"`
	Entry           string    `json:"entry" db:"entry"`
	DisplayOnGarden bool      `json:"display_on_garden" db:"display_on_garden"`
	Image           string    `json:"image" db:"image"`
	Category        Category  `json:"category" db:"category"`
	PlantID         uuid.UUID `json:"-" db:"plant_id"`
	Plant           *Plant    `json:"Plant,omitempty" belongs_to:"plant"`
	CreatedAt       time.Time `json:"created_at" db:"created_at"`
	UpdatedAt       time.Time `json:"updated_at" db:"updated_at"`
}

// String is not required by pop and may be deleted
func (j Journal) String() string {
	jj, _ := json.Marshal(j)
	return string(jj)
}

func (c Category) CatStr() string {
	return string(c)
}

// watch this ticket https://github.com/gobuffalo/fizz/issues/65, fizz does not support enums yet
func IsValidCategory(category string) bool {
	validCategories := []Category{Pests, Planting, Watering, Pruning, Harvesting, Weather, Germination}

	for _, v := range validCategories {
		if string(v) == category {
			return true
		}
	}
	return false
}

// Journals is not required by pop and may be deleted
type Journals []Journal

// String is not required by pop and may be deleted
func (j Journals) String() string {
	jj, _ := json.Marshal(j)
	return string(jj)
}

// Validate gets run every time you call a "pop.Validate*" (pop.ValidateAndSave, pop.ValidateAndCreate, pop.ValidateAndUpdate) method.
// This method is not required and may be deleted.
func (j *Journal) Validate(tx *pop.Connection) (*validate.Errors, error) {
	return validate.Validate(
		&validators.StringIsPresent{Field: j.Title, Name: "Title"},
		&validators.StringIsPresent{Field: j.Entry, Name: "Entry"},
		&validators.StringIsPresent{Field: j.Category.CatStr(), Name: "Category"},
	), nil
}

// ValidateCreate gets run every time you call "pop.ValidateAndCreate" method.
// This method is not required and may be deleted.
func (j *Journal) ValidateCreate(tx *pop.Connection) (*validate.Errors, error) {
	if !IsValidCategory(string(j.Category)) {
		vErrs := validate.NewErrors()
		vErrs.Add("category", "invalid category")
		return vErrs, nil
	}
	return validate.NewErrors(), nil
}

// ValidateUpdate gets run every time you call "pop.ValidateAndUpdate" method.
// This method is not required and may be deleted.
func (j *Journal) ValidateUpdate(tx *pop.Connection) (*validate.Errors, error) {
	if !IsValidCategory(string(j.Category)) {
		vErrs := validate.NewErrors()
		vErrs.Add("category", "invalid category")
		return vErrs, nil
	}
	return validate.NewErrors(), nil
}

File: frontend/lib/model/journal.dart
class Journal {
  final String id;
  final String title;
  final String entry;
  final bool display_on_garden;
  final String? image;
  final String category;
  final String plant_id;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int marked_for_deletion;

  Journal(
      {required this.id,
      required this.title,
      required this.entry,
      required this.display_on_garden,
      this.image,
      required this.category,
      required this.plant_id,
      required this.createdAt,
      required this.updatedAt,
      this.marked_for_deletion = 0});

  factory Journal.fromJson(Map<String, dynamic> json) {
    bool parseBool(dynamic value) {
      if (value is bool) return value;
      if (value is int) return value == 1;
      throw ArgumentError('Invalid boolean value: $value');
    }

    return Journal(
      id: json['id'],
      title: json['title'],
      entry: json['entry'],
      display_on_garden: parseBool(json['display_on_garden']),
      image: json['image'],
      category: json['category'],
      plant_id: json['plant_id'],
      marked_for_deletion: json['marked_for_deletion'] ?? 0,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : DateTime.now(),
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson({bool forSqlLite = false}) {
    dynamic serializeBool(bool value) {
      return forSqlLite ? (value ? 1 : 0) : (value ? 't' : 'f');
    }

    return {
      'id': id,
      'title': title,
      'entry': entry,
      'display_on_garden': serializeBool(display_on_garden),
      'image': image,
      'category': category,
      'plant_id': plant_id,
      'marked_for_deletion': marked_for_deletion,
      'created_at': createdAt.toUtc().toIso8601String(),
      'updated_at': updatedAt.toUtc().toIso8601String(),
    };
  }
}
